temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java

```diff
@@ -86,22 +92,50 @@ public WorkflowStartOutput start(WorkflowStartInput input) {
                 .build()
             : null;
 
-    StartWorkflowExecutionRequest.Builder request =
+    StartWorkflowExecutionRequest.Builder startRequest =
         requestsHelper.newStartWorkflowExecutionRequest(
             input.getWorkflowId(),
             input.getWorkflowType(),
             input.getHeader(),
             input.getOptions(),
             inputArgs.orElse(null),
             memo);
+
+    StartWorkflowOperation startOperation = input.getOptions().getStartOperation();
+    if (startOperation != null) {
+      if (startOperation instanceof WorkflowStartOperationUpdate) {
+        WorkflowStartOperationUpdate<?> updateOperation =
+            (WorkflowStartOperationUpdate<?>) startOperation;
+
+        if (!updateOperation.getWorkflowType().isEmpty()
+            && !updateOperation.getWorkflowType().equals(input.getWorkflowType())) {
+          throw new IllegalArgumentException(
+              "Update Operation does not belong to '"
+                  + input.getWorkflowType()
+                  + "' but '"
+                  + updateOperation.getWorkflowType()
+                  + "'");
+        }
```

Comment:
This is not 100% bullet proof but might be good enough. I couldn't get to a place where I had the Update and the Start's full Java class available for comparison. Using the workflow type seemed like the next best thing.